### ISSUE TITLE
[Feature Request] data class like behaviour

### ISSUE THREAD

**Issue by vmoens** [pytorch-labs/tensordict#102]
## Motivation

Some users have pointed out that data classes offer advantages over dict-like classes:
- predictability: one knows in advance what fields to expect, the code is transparent although less generic
- autocomplete: IDEs can keep track of the attributes through type hinting 
- dedicated behaviour: a class can have a set of dedicated methods that act on its attributes.

## Proposed API

We could have a mixed approach where a data class would build a tensordict and interface with it. Attributes would point to keys. The methods would be shared across classes.

```python
@tensordict_data  # better name?
class MyData:
   float_tensor
   sparse_tensor

   def stuff(self):
       return self.float_tensor + self.sparse_tensor 

data = MyData(float_tensor=a, sparse_tensor=b, batch_size=[3, 4])
data.reshape(-1) # returns a new MyData with shape/ batch_size of [12]
data.batch_size # picks the batch size of the tensordict
data.missing_key # returns a dedicated error since this key is not expected
```

## Challenges
Any such dataclass methods (such as reshape, split etc) should return a new instance of the same class. 
Upon creation (at loading time) one should check that the data class attributes do not collide with the tensordict ones.
Nesting with data classes will be slightly harder than with tensordict (see [this thread](https://www.linkedin.com/feed/update/urn:li:activity:7003878068733779968?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A7003878068733779968%2C7004083519140323328%29))

**Comment by roccajoseph** [comment]
I started working on it.

**Comment by tcbegley** [comment]
Resolved by #107
